src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1347 @@
+
+Author: Romain Brenguier
+
+Date:   September 2016
+
+\*******************************************************************/
+
+#include "string_refine_preprocess.h"
+
+#include <util/std_expr.h>
+#include <util/symbol_table.h>
+#include <util/pointer_offset_size.h>
+#include <util/prefix.h>
+#include <solvers/refinement/string_expr.h>
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::new_tmp_symbol
+
+  Inputs: a name and a type
+
+ Outputs: a new symbol
+
+ Purpose: add a temporary symbol with the given name and type to the
+          symbol table and returns it
+
+\*******************************************************************/
+
+symbol_exprt string_refine_preprocesst::new_tmp_symbol(
+  const std::string &name, const typet &type)
+{
+  auxiliary_symbolt tmp_symbol;
+  // TODO: add a prefix to the name to avoid clashing with other
+  // variables of the program
+  tmp_symbol.base_name=name;
+  tmp_symbol.is_static_lifetime=false;
+  tmp_symbol.mode=ID_java;
+  tmp_symbol.name=name;
+  tmp_symbol.type=type;
+  symbol_table.add(tmp_symbol);
+  return symbol_exprt(name, type);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::new_tmp_symbol
+
+  Inputs: a name and a type
+
+ Outputs: a new symbol
+
+ Purpose: add a symbol with the given name and type to the symbol table
+          and returns it
+
+\*******************************************************************/
+
+symbol_exprt string_refine_preprocesst::new_symbol(
+  const std::string &prefix, const typet &type, bool static_lifetime)
+{
+  symbolt tmp_symbol;
+  // TODO: add a prefix to the name to avoid clashing with other
+  // variables of the program
+  std::ostringstream buf;
+  buf << prefix << "_" << (next_symbol_id++);
+  std::string name = buf.str();
+  tmp_symbol.base_name=name;
+  tmp_symbol.is_static_lifetime=static_lifetime;
+  tmp_symbol.is_state_var=static_lifetime;
+  tmp_symbol.is_type=false;
+  tmp_symbol.is_lvalue=static_lifetime; // not sure about that
+  tmp_symbol.mode=ID_java;
+  tmp_symbol.name=name;
+  tmp_symbol.type=type;
+  symbol_table.add(tmp_symbol);
+  return symbol_exprt(name,type);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::declare_function
+
+  Inputs: a name and a type
+
+ Purpose: declare a function with the given name and type
+
+\*******************************************************************/
+
+void string_refine_preprocesst::declare_function(
+  irep_idt function_name, const typet &type)
+{
+  auxiliary_symbolt func_symbol;
+  func_symbol.base_name=function_name;
+  func_symbol.is_static_lifetime=false;
+  func_symbol.mode=ID_java;
+  func_symbol.name=function_name;
+  func_symbol.type=type;
+  symbol_table.add(func_symbol);
+  goto_functions.function_map[function_name];
+}
+
+static bool is_java_string_type(const typet& type)
```

Comment:
Actually it is already declared here https://github.com/diffblue/cbmc/blob/master/src/solvers/refinement/refined_string_type.h#L64 so I'm going to use this one for now, but yes it could be moved